google-cloud-datastore/src/main/java/com/google/cloud/datastore/spi/v1/HttpDatastoreRpc.java

```diff
@@ -41,174 +44,103 @@
-        censusHttpModule.getHttpRequestInitializer(
-            httpTransportOptions.getHttpRequestInitializer(options));
-
-    final String applicationName = options.getApplicationName();
-    return new HttpRequestInitializer() {
-      @Override
-      public void initialize(HttpRequest httpRequest) throws IOException {
-        censusHttpModuleHttpRequestInitializer.initialize(httpRequest);
-        httpRequest.getHeaders().setUserAgent(applicationName);
-      }
-    };
-  }
+public class HttpDatastoreRpc implements AutoCloseable, DatastoreRpc {
 
-  private static boolean isLocalHost(String host) {
-    if (!host.isEmpty()) {
-      try {
-        String normalizedHost = "http://" + removeScheme(host);
-        InetAddress hostAddr = InetAddress.getByName(new URL(normalizedHost).getHost());
-        return hostAddr.isAnyLocalAddress() || hostAddr.isLoopbackAddress();
-      } catch (Exception e) {
-        // ignore
-      }
-    }
-    return false;
-  }
+  private final ClientContext clientContext;
+  private final HttpJsonDatastoreStub datastoreStub;
 
-  private static String removeScheme(String url) {
-    if (url != null) {
-      if (url.startsWith("https://")) {
-        return url.substring("https://".length());
-      } else if (url.startsWith("http://")) {
-        return url.substring("http://".length());
-      }
-    }
-    return url;
-  }
+  private boolean closed;
 
-  private static DatastoreException translate(
-      com.google.datastore.v1.client.DatastoreException exception) {
-    return translate(exception, true);
-  }
+  public HttpDatastoreRpc(DatastoreOptions options) throws IOException {
+    DatastoreSettings datastoreSettings =
+        new DatastoreSettingsBuilder(DatastoreSettings.newBuilder().build())
+            .setInternalHeaderProvider(
+                DatastoreStubSettings.defaultHttpJsonApiClientHeaderProviderBuilder().build())
+            .setTransportChannelProvider(
+                DatastoreStubSettings.defaultHttpJsonTransportProviderBuilder().build())
+            .setEndpoint(getHost(options))
+            .build();
 
-  private static DatastoreException translate(
-      com.google.datastore.v1.client.DatastoreException exception, boolean idempotent) {
-    String reason = "";
-    if (exception.getCode() != null) {
-      reason = exception.getCode().name();
-    }
-    if (reason.isEmpty()) {
-      if (exception.getCause() instanceof IOException) {
-        return new DatastoreException((IOException) exception.getCause());
-      }
-    }
-    return new DatastoreException(
-        exception.getCode().getNumber(), exception.getMessage(), reason, idempotent, exception);
+    clientContext = ClientContext.create(datastoreSettings);
+
+    ApiFunction<UnaryCallSettings.Builder<?, ?>, Void> retrySettingsSetter =
+        builder -> {
+          builder.setRetrySettings(options.getRetrySettings());
+          return null;
+        };
+    DatastoreStubSettings datastoreStubSettings =
```

Comment:
it would be good to add some testing around this (and grpc impl) that retry settings are as expected

Reply:
i've moved that to a common place i.e. `DatastoreRpc` and was able to test some of it.